scrapy/squeues.py

```diff
@@ -2,20 +2,28 @@
 Scheduler queues
 """
 
+from __future__ import annotations
+
 import marshal
 import pickle  # nosec
 from os import PathLike
 from pathlib import Path
-from typing import Union
+from typing import TYPE_CHECKING, Any, Callable, Optional, Type, Union
 
 from queuelib import queue
 
+from scrapy import Request
+from scrapy.crawler import Crawler
 from scrapy.utils.request import request_from_dict
 
+if TYPE_CHECKING:
+    # typing.Self requires Python 3.11
+    from typing_extensions import Self
+
 
-def _with_mkdir(queue_class):
-    class DirectoriesCreated(queue_class):
-        def __init__(self, path: Union[str, PathLike], *args, **kwargs):
+def _with_mkdir(queue_class: Type[queue.BaseQueue]) -> Type[queue.BaseQueue]:
+    class DirectoriesCreated(queue_class):  # type: ignore[valid-type,misc]
```

Comment:
Still not sure how to do this correctly, and whether this is a deficiency in mypy, something the hints can't express or just wrong hints.